tests/test-amp-image-dimension-extractor.php

```diff
@@ -49,92 +51,160 @@ function test__normalize_url( $source_url, $expected_url ) {
+		$sources = array_fill_keys( $sources, false );
+		$dimensions = AMP_Image_Dimension_Extractor::extract_by_downloading_images( $sources );
 
-		$dimensions = AMP_Image_Dimension_Extractor::extract_from_attachment_metadata(  $source_dimensions, $source );
 		$this->assertEquals( $expected, $dimensions );
 	}
 
-	function test__invalid_attachment() {
-		$source = 'https://example.com/path/to/file.jpg';
-		$expected = false;
+	function test__valid_image_file_synchronous() {
+		$sources = array(
+			IMG_350,
+		);
+		$expected = array(
+			IMG_350 => array(
+				'width' => 350,
+				'height' => 150,
+			),
+		);
 
-		$dimensions = AMP_Image_Dimension_Extractor::extract_from_attachment_metadata(  false, $source );
+		$sources = array_fill_keys( $sources, false );
+		$dimensions = AMP_Image_Dimension_Extractor::extract_by_downloading_images( $sources, 'synchronous' );
 
 		$this->assertEquals( $expected, $dimensions );
 	}
 
-	function test__valid_attachment_missing_metadata() {
-		$source = wp_get_attachment_url( $this->_attachment_id );
-		$expected = false;
+	function test__multiple_valid_image_files() {
+		$sources = array(
+			IMG_350,
+			IMG_1024,
+		);
+		$expected = array(
+			IMG_350 => array(
+				'width' => 350,
+				'height' => 150,
+			),
+			IMG_1024 => array(
+				'width' => 1024,
+				'height' => 768,
+			),
+		);
 
-		delete_post_meta( $this->_attachment_id, '_wp_attachment_metadata' );
-		$dimensions = AMP_Image_Dimension_Extractor::extract_from_attachment_metadata( false, $source );
+		$sources = array_fill_keys( $sources, false );
+		$dimensions = AMP_Image_Dimension_Extractor::extract_by_downloading_images( $sources );
 
 		$this->assertEquals( $expected, $dimensions );
 	}
 
-	function test__valid_attachment_and_dimensions_with_query() {
-		$source = wp_get_attachment_url( $this->_attachment_id ) . '?rand=1';
-		$expected = array( 498, 113 );
+	function test__multiple_valid_image_files_synchronous() {
+		$sources = array(
+			IMG_350,
+			IMG_1024,
+		);
+		$expected = array(
+			IMG_350 => array(
+				'width' => 350,
+				'height' => 150,
+			),
+			IMG_1024 => array(
+				'width' => 1024,
+				'height' => 768,
+			),
+		);
 
-		$dimensions = AMP_Image_Dimension_Extractor::extract_from_attachment_metadata( false, $source );
+		$sources = array_fill_keys( $sources, false );
+		$dimensions = AMP_Image_Dimension_Extractor::extract_by_downloading_images( $sources, 'synchronous' );
 
 		$this->assertEquals( $expected, $dimensions );
 	}
 
-	function test__valid_attachment_and_dimensions() {
-		$source = wp_get_attachment_url( $this->_attachment_id );
-		$expected = array( 498, 113 );
+	function test__invalid_image_file() {
+		$sources = array(
+			AMP_IMG_DIMENSION_TEST_INVALID_FILE,
+		);
+		$expected = array(
+			AMP_IMG_DIMENSION_TEST_INVALID_FILE => false,
+		);
 
-		$dimensions = AMP_Image_Dimension_Extractor::extract_from_attachment_metadata( false, $source );
+		$sources = array_fill_keys( $sources, false );
```

Comment:
Rather than calling `array_fill_keys` for the test cases, we should just set up the `sources` variables with the data format that the `extract_by_*` method expects (i.e. remove as much logic from tests as possible).